examples/autoignition-mpi.py

```diff
@@ -240,71 +309,140 @@ def main(ctx_factory=cl.create_some_context, casename="autoignition", use_leap=F
                 "global_nelements": global_nelements,
                 "num_parts": nproc
             }
             from mirgecom.restart import write_restart_file
-            write_restart_file(actx, rst_data, rst_filename, comm)
-
-        # awful - computes potentially expensive viz quantities
-        #         regardless of whether it is time to viz
-        reaction_rates = eos.get_production_rates(state)
-        viz_fields = [("reaction_rates", reaction_rates)]
-        return sim_checkpoint(discr, visualizer, eos, cv=state,
-                              vizname=casename, step=step,
-                              t=t, dt=dt, nstatus=nstatus, nviz=nviz,
-                              constant_cfl=constant_cfl, comm=comm,
-                              viz_fields=viz_fields)
-
-    try:
-        (current_step, current_t, current_state) = \
-            advance_state(rhs=my_rhs, timestepper=timestepper,
-                checkpoint=my_checkpoint,
-                get_timestep=get_timestep, state=current_state,
-                t=current_t, t_final=t_final)
-    except ExactSolutionMismatch as ex:
-        error_state = True
-        current_step = ex.step
-        current_t = ex.t
-        current_state = ex.state
-
-    if not check_step(current_step, nviz):  # If final step not an output step
-        if rank == 0:
-            logger.info("Checkpointing final state ...")
-        my_checkpoint(current_step, t=current_t,
-                      dt=(current_t - checkpoint_t),
-                      state=current_state)
-
-    if current_t - t_final < 0:
-        error_state = True
-
-    if error_state:
-        raise ValueError("Simulation did not complete successfully.")
+            write_restart_file(actx, rst_data, rst_fname, comm)
+
+    def my_health_check(dv):
+        health_error = False
+        from mirgecom.simutil import check_naninf_local, check_range_local
+        if check_naninf_local(discr, "vol", dv.pressure) \
+           or check_range_local(discr, "vol", dv.pressure, 1e5, 2.4e5):
+            health_error = True
+            logger.info(f"{rank=}: Invalid pressure data found.")
+
+        if check_range_local(discr, "vol", dv.temperature, 1.4e3, 3.3e3):
+            health_error = True
+            logger.info(f"{rank=}: Invalid temperature data found.")
+
+        return health_error
+
+    def my_pre_step(step, t, dt, state):
+        try:
+            dv = None
+
+            if logmgr:
+                logmgr.tick_before()
+
+            from mirgecom.simutil import check_step
+            do_viz = check_step(step=step, interval=nviz)
+            do_restart = check_step(step=step, interval=nrestart)
+            do_health = check_step(step=step, interval=nhealth)
+
+            if do_health:
+                dv = eos.dependent_vars(state)
+                health_errors = my_health_check(dv)
+                if comm is not None:
+                    health_errors = comm.allreduce(health_errors, op=MPI.LOR)
```

Comment:
Think it's worth adding in a utility function in `simutils` that does this reduction across ranks? Just to clean this up. I think @majosm has some code somewhere that does this. This is not a functional request, just aesthetics.

Reply:
`health_errors = allsync(my_health_check(dv), op=MPI.LOR)` ?

Reply:
I mean, `comm` is already being used around so is it _that_ bad pass it? I don't see the problem.

Reply:
Here is the code snippet @thomasgibson was referring to: https://github.com/illinois-ceesd/mirgecom/blob/e47d3d0b39b4782c05ef3d27d8cc63ceb84c47a3/mirgecom/mpi.py#L161-L166

(Don't copy the implementation, it's dumb. Just use `MPI.LOR`.)

Reply:
Not a problem at all!

Reply:
Just real quick before I implement..... @thomasgibson and @majosm 

How would we feel about this utility:

```python
def allsync(local_value(s), comm, op=MPI.MAX):
   if comm is None:
     return local_value(s)
   return comm.allreduce(local_values, op=op)
 
global_value(s) = allsync(local_value(s), comm, op=MPIOP)
```

This way the user can use the `allsync` utility to do syncing operations on arrays of data, or other objects as supported by comm, and control over what operation is performed.